File: ts/indicators/common/simple-moving-average/simple-moving-average.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { SimpleMovingAverageOutput } from './simple-moving-average-output';
import { SimpleMovingAverageParams } from './simple-moving-average-params';

/** Function to calculate mnemonic of a __SimpleMovingAverage__ indicator. */
export const simpleMovingAverageMnemonic = (params: SimpleMovingAverageParams): string =>
  'sma('.concat(params.length.toString(), componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent), ')');

/** Simple Moving Average line indicator. */
export class SimpleMovingAverage extends LineIndicator {
  private window: Array<number>;
  private windowLength: number;
  private windowSum: number;
  private windowCount: number;
  private lastIndex: number;

  /**
   * Constructs an instance given a length in samples.
   * The length should be an integer greater than 1.
   **/
  public constructor(params: SimpleMovingAverageParams){
    super();
    const length = Math.floor(params.length);
    if (length < 2) {
      throw new Error('length should be greater than 1');
    }

    this.mnemonic = simpleMovingAverageMnemonic(params);
    this.description = 'Simple moving average ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;
    this.window = new Array<number>(length);
    this.windowLength = length;
    this.windowSum = 0;
    this.windowCount = 0;
    this.lastIndex = length - 1;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.SimpleMovingAverage,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: SimpleMovingAverageOutput.SimpleMovingAverageValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    if (this.primed) {
      this.windowSum += sample - this.window[0];
      for (let i = 0; i < this.lastIndex; i++) {
        this.window[i] = this.window[i+1];
      }

      this.window[this.lastIndex] = sample;
    } else {
      this.windowSum += sample;
      this.window[this.windowCount] = sample;
      this.windowCount++;

      if (this.windowLength > this.windowCount) {
        return Number.NaN;
      }

      this.primed = true;
    }

    return this.windowSum / this.windowLength;
  }
}

File: go/indicators/common/simplemovingaverage/simplemovingaverage.go
package simplemovingaverage

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// SimpleMovingAverage computes the simple, or arithmetic, moving average (SMA) by adding the samples
// for a number of time periods (length, ℓ) and then dividing this total by the number of time periods.
//
// In other words, this is an unweighted mean (gives equal weight to each sample) of the previous ℓ samples.
//
// This implementation updates the value of the SMA incrementally using the formula:
//
//	SMAᵢ = SMAᵢ₋₁ + (Pᵢ - Pᵢ₋ℓ) / ℓ,
//
// where ℓ is the length.
//
// The indicator is not primed during the first ℓ-1 updates.
type SimpleMovingAverage struct {
	mu sync.RWMutex
	core.LineIndicator
	window       []float64
	windowSum    float64
	windowLength int
	windowCount  int
	lastIndex    int
	primed       bool
}

// NewSimpleMovingAverage returns an instnce of the indicator created using supplied parameters.
func NewSimpleMovingAverage(p *SimpleMovingAverageParams) (*SimpleMovingAverage, error) {
	const (
		invalid = "invalid simple moving average parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "sma(%d%s)"
		minlen  = 2
	)

	length := p.Length
	if length < minlen {
		return nil, fmt.Errorf(fmts, invalid, "length should be greater than 1")
	}

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Build mnemonic using resolved components — defaults are omitted by ComponentTripleMnemonic.
	mnemonic := fmt.Sprintf(fmtn, length, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Simple moving average " + mnemonic

	sma := &SimpleMovingAverage{
		window:       make([]float64, length),
		windowLength: length,
		lastIndex:    length - 1,
	}

	sma.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, sma.Update)

	return sma, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *SimpleMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
// It always has a single scalar output -- the calculated value of the simple moving average.
func (s *SimpleMovingAverage) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.SimpleMovingAverage,
		Mnemonic:    s.LineIndicator.Mnemonic,
		Description: s.LineIndicator.Description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(SimpleMovingAverageValue),
				Type:        outputs.ScalarType,
				Mnemonic:    s.LineIndicator.Mnemonic,
				Description: s.LineIndicator.Description,
			},
		},
	}
}

// Update updates the value of the simple moving average given the next sample.
//
// The indicator is not primed during the first ℓ-1 updates.
func (s *SimpleMovingAverage) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	temp := sample

	s.mu.Lock()
	defer s.mu.Unlock()

	if s.primed {
		s.windowSum += temp - s.window[0]

		for i := 0; i < s.lastIndex; i++ {
			s.window[i] = s.window[i+1]
		}

		s.window[s.lastIndex] = temp
	} else {
		s.windowSum += temp
		s.window[s.windowCount] = temp
		s.windowCount++

		if s.windowLength > s.windowCount {
			return math.NaN()
		}

		s.primed = true
	}

	return s.windowSum / float64(s.windowLength)
}
